MAILPOET-5496/update-used-shipping-and-payment-methods-timespan

johnolek opened the pull request:
## Description

This adds the "over all time" option to the used shipping method and used payment method dynamic segment filters, which we missed when working on MAILPOET-4991 (https://github.com/mailpoet/mailpoet/pull/5057).

## Code review notes

_N/A_

## QA notes

This update changes how the data for the number of days is stored for these two filters, and it includes a migration to handle that. Please make sure that filters created before this version continue to work exactly as they did before after the update, and please also make sure that the migration ran successfully before testing (I think it should just require deactivating and reactivating the plugin).

## Linked PRs

https://github.com/mailpoet/mailpoet/pull/5057 completed this work for other fliters.
I used some of the changes introduced in https://github.com/mailpoet/mailpoet/pull/5064, so that will need to be merged before this will be ready to review (edit: that PR has been merged and this PR has been rebased)

## Linked tickets
[MAILPOET-5496](https://mailpoet.atlassian.net/browse/MAILPOET-5496)

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5496]: https://mailpoet.atlassian.net/browse/MAILPOET-5496?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
Add lifetime option to payment/shipping filters

MAILPOET-5496

Commit:
Migrate existing shipping/payment method filter data

Now that these filters are using the shared DaysPeriodField, they store
the day count in a field named days instead of used_payment_method_days
or used_shipping_method_days.

This migration is based on Migration_20230712_180341, where we needed to
 do the same thing for other filters.

MAILPOET-5496

rodrigoprimo commented on mailpoet/lib/Segments/DynamicSegments/Filters/WooCommerceUsedPaymentMethod.php:
```diff
@@ -52,27 +53,30 @@ public function apply(QueryBuilder $queryBuilder, DynamicSegmentFilterEntity $fi
-      throw new InvalidFilterException('Missing days', InvalidFilterException::MISSING_VALUE);
+    $data = $filterData->getData();
+    // Backwards compatibility for filters saved before timeframe was added
+    if (!isset($data['timeframe'])) {
```
Have you considered adding the `timeframe` in the migration instead of having this backward compatibility code? Maybe it is better to have it here in case this code runs before the migration is executed?

johnolek replied:
Thanks for this suggestion! I think adding it to the migrationand removing the backwards compatibility code makes a lot more sense. I've updated the migration in 7b12046ceec7eaefa9ddb071a833398e70ca48b8. Can you think of any way that the new migration could potentially fail? The update is wrapped in some conditionals, but they should always be true for any saved filters.

rodrigoprimo replied:
Not really, I just assumed that was why you opted to add the BC code. Creating the `timeframe` in the migration seems like a good approach to me.

rodrigoprimo reviewed (commented):
Good job! Looks good to me. I just left a question for us to discuss and to confirm something that I'm assuming. Probably no change will be necessary.

Commit:
Include timeframe in migration to avoid BC code

MAILPOET-5496

johnolek commented:
FYI the last force push was just rebasing this branch on `trunk`, it didn't include any changes since the review.

Commit:
Ensure shipping method instanceIds are strings

Selected shipping methods were not appearing when after saving and
reloading a segment because the instanceIds were actually coming through
 as integers, while the values are getting saved as strings.

MAILPOET-5496

Aschepikov reviewed (approved):
GJ! Thanks for the fixing
